cpp/ql/test/experimental/query-tests/Security/CWE/CWE-401/semmle/tests/test.c

```diff
@@ -0,0 +1,108 @@
+#define size_t int
+#define NULL ((void*)0)
+
+
+unsigned char * badResize0(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	// BAD: on unsuccessful call to realloc, we will lose a pointer to a valid memory block
+	if (currentSize < newSize)
+	{
+		buffer = (unsigned char *)realloc(buffer, newSize);
+	}
+	return buffer;
+}
+
+unsigned char * goodResize0(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	// GOOD: this way we will exclude possible memory leak 
+	unsigned char * tmp;
+	if (currentSize < newSize)
+	{
+		tmp = (unsigned char *)realloc(buffer, newSize);
+	}
+	if (tmp == NULL)
+	{
+		free(buffer);
+		return NULL;
+	} 
+	else
+		buffer = tmp;
+	return buffer;
+}
+unsigned char * badResize1(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	if(!buffer)
+		exit(0);
+	// BAD: on unsuccessful call to realloc, we will lose a pointer to a valid memory block
+	if (currentSize < newSize)
+	{
+		buffer = (unsigned char *)realloc(buffer, newSize);
+	}
+	return buffer;
+}
+
+unsigned char * noBadResize1(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	// GOOD: program to end
+	if (currentSize < newSize)
+	{
+		buffer = (unsigned char *)realloc(buffer, newSize);
+	}
+	if(!buffer)
+		exit(0);
+	return buffer;
+}
+unsigned char * noBadResize1e(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	// GOOD: program to end
+	if (currentSize < newSize)
+	{
+		buffer = (unsigned char *)realloc(buffer, newSize);
+	}
+	if(buffer)
+		return buffer;
+	else
+		exit(0);
+}
+unsigned char * noBadResize1o(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	// GOOD: program to end
+	if (currentSize < newSize)
+	{
+		if(buffer = (unsigned char *)realloc(buffer, newSize))
+			exit(0);
+	}
+	return buffer;
+}
+unsigned char * badResize2(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	assert(buffer!=0);
+	// BAD: on unsuccessful call to realloc, we will lose a pointer to a valid memory block
+	if (currentSize < newSize)
+	{
+		buffer = (unsigned char *)realloc(buffer, newSize);
+	}
+	return buffer;
+}
+
+unsigned char * noBadResize2(unsigned char * buffer,size_t currentSize,size_t newSize)
+{
+	// GOOD: program to end
+	if (currentSize < newSize)
+	{
+		buffer = (unsigned char *)realloc(buffer, newSize);
+		assert(buffer!=0);
```

Comment:
please clarify here, you mean fix the test for assert (buffer); or talk about extending the test with different variations of assert calls. sorry for the question, I'm just not a native speaker. (

Reply:
I suggest adding code for `assert` into your test.  Something like:
```
#define assert(x) if (!(x)) assert_fail(__FILE__,__LINE__)
void assert_fail(const char *file, int line) { exit(1); }
```
That way it will be clear to the reader, and to the analysis, what `assert` actually means.